ng-dev/utils/yargs.ts

```diff
@@ -0,0 +1,26 @@
+import yargs, {Argv} from 'yargs';
+
+// A function to be called when the command completes.
+type CompletedFn = (err: Error | null) => Promise<void> | void;
+
+/** List of functions to be called upon command completion. */
+const completedFunctions: CompletedFn[] = [];
+
+/** Register a function to be called when the command completes. */
+export function registerCompletedFunction(fn: CompletedFn) {
+  completedFunctions.push(fn);
+}
+
+/**
+ * Run the yargs process, as configured by the supplied function, calling a set of completion
+ * functions after the command completes.
+ */
+export function runParserWithCompletedFunctions(applyConfiguration: (argv: Argv) => Argv) {
+  applyConfiguration(yargs(process.argv.slice(2)))
+    .exitProcess(false)
+    .parse(process.argv.slice(2), async (err: Error | null) => {
```

Comment:
In the docs there is also an `output` parameter. It sounds like this one contains necessary information that would otherwise be printed? I think we need to be careful about not causing unexpected scenarios here.

Have you considered just keeping the register approach (I like that), but then just subscribing to `process.on('exit'` or something?

Also what about the exit code if `err` is not `null`?

Reply:
`output` contains the string that has been printed to the console throughout the command, so we don't need its information.

The problem using `process.on('exit'` is that if something prevents the `exit` from occurring then we don't every run the completed functions.  We want to tie the completed functions to the completion of the command specifically.

The error would have been intentionally thrown, but by providing it to the completed functions, functions can act upon the error that was thrown.

Reply:
I don't know if Yarn could capture our log messages as it doesn't patch e.g. console.log.

Quote from the docs

> output: any text that would have been output by yargs to the terminal, had a callback not been provided

Sounds like this is something different? 

Makes sense with the explanation on `process.exit`. Good point.

Reply:
I see what you are saying, I think I tested that and then ended up switching things around.  Updated.

Reply:
I read more up and it seems like the most idiomatic solution would really be to just use a top level async/await and try/catch/finally.

https://github.com/yargs/yargs/pull/1823

Let's go with this 😆 but we can clean this up as it seems. I would like the more obvious/predictable top level await more